src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala

```diff
@@ -0,0 +1,174 @@
+        startOffset.store(region.allocate(t.contentsByteSize(length)))
+    )
+    if (pOffset != null) {
+      c = Code(c, region.storeAddress(pOffset, startOffset))
+    }
+    if (init)
+      c = Code(c, t.initialize(region, startOffset.load(), length, idx))
+    c = Code(c,
+      (length < 10).mux(
+        elementsOffset.store(startOffset.load() + elementsOffsetArray(length)),
+        elementsOffset.store(startOffset.load() + t.elementsOffset(length))
+      )
+    )
+    Code(c, idx.store(0))
+  }
+
+  def start(init: Boolean): Code[Unit] = {
+    val t = rowType.asInstanceOf[TStruct]
+    var c = if (pOffset == null)
+      Code(
+        region.align(t.alignment),
+        startOffset.store(region.allocate(t.byteSize))
+      )
+    else
+      startOffset.store(pOffset)
+    c = Code(c, elementsOffset.store(startOffset + t.byteOffsets(0)))
+    if (init)
+      c = Code(c,t.clearMissingBits(region, startOffset))
+    c
+  }
+
+  def setMissing(): Code[Unit] = {
+    rowType match {
+      case t: TArray => t.setElementMissing(region, startOffset, idx)
+      case t: TStruct => t.setFieldMissing(region, startOffset, staticIdx)
+    }
+  }
+
+  def addInt32(v: Code[Int]): Code[Unit] = region.storeInt32(currentOffset, v)
+
+  def addInt64(v: Code[Long]): Code[Unit] = region.storeInt64(currentOffset, v)
+
+  def addFloat32(v: Code[Float]): Code[Unit] = region.storeFloat32(currentOffset, v)
+
+  def addFloat64(v: Code[Double]): Code[Unit] = region.storeFloat64(currentOffset, v)
+
+  def addBinary(bytes: Code[Array[Byte]]): Code[Unit] = {
+    Code(
+      region.align(TBinary.contentAlignment),
+      rowType.fundamentalType match {
+        case TBinary => _empty
+        case _ =>
+          region.storeAddress(currentOffset,endOffset)
+      },
+      region.appendInt32(bytes.length()),
+      region.appendBytes(bytes)
+    )
+  }
+
+  def addString(str: Code[String]): Code[Unit] = addBinary(str.invoke[Array[Byte]]("getBytes"))
+
+  def addArray(t: TArray, f: (StagedRegionValueBuilder[T] => Code[Unit])): Code[Unit] = f(new StagedRegionValueBuilder[T](fb, t, this))
+
+  def addStruct(t: TStruct, f: (StagedRegionValueBuilder[T] => Code[Unit]), init: LocalRef[Boolean] = null): Code[Unit] = f(new StagedRegionValueBuilder[T](fb, t, this))
+
+  def advance(): Code[Unit] = {
+    rowType match {
+      case t: TArray => Code(
+        elementsOffset := elementsOffset + t.elementByteSize,
+        new Code[Unit] {
+          def emit(il: Growable[AbstractInsnNode]):Unit = {
+            il += new IincInsnNode(idx.i, 1)
+          }
+      }
+      )
+      case t: TStruct => {
+        staticIdx += 1
+        if (staticIdx < t.size)
+          elementsOffset.store(startOffset + t.byteOffsets(staticIdx))
+        else _empty
+      }
+      case _ => _empty
+    }
+  }
+
+  def build() {
+    emit(_return(startOffset))
+    transform = fb.result()
+  }
+
+  def emit(c: Code[_]) {fb.emit(c)}
+
+  def emit(cs: Array[Code[_]]) {
```

Comment:
I think it's better to keep these kinds of functionality on the `FunctionBuilder` only.